### PR TITLE
Switch from logrus to k8s.io/klog

### LINKED ISSUE #109: consider using klog for logging
I think https://github.com/kubernetes/klog is the way forward for the ecosystem, and after some experimenting I think structured logging makes the most sense for server / syslog type logs, not debugging a CLI.

I don't love glog/klogs's line formatting, but it seems like the way forward to integrate in the ecosystem.

cc @munnerz

### PR DESCRIPTION
**Pull request by alejandrox1**
Resolves #109

Signed-off-by: Jorge Alarcon Ochoa <alarcj137@gmail.com>

### CODE REVIEW

File: cmd/kind/kind.go [hunk 1]
```diff
@@ -68,14 +57,6 @@ func NewCommand() *cobra.Command {
 }
 
 func runE(flags *Flags, cmd *cobra.Command, args []string) error {
-	level := defaultLevel
```
Comment (BenTheElder) [inline cmd/kind/kind.go]: We should have a new flag for the klog level

File: pkg/log/status.go [hunk 2]
```diff
@@ -120,12 +114,6 @@ func (s *Status) End(success bool) {
 		return
 	}
 
-	isTerm := IsTerminal(s.writer)
```
[the remark this answers is not in the document]
Comment (alejandrox1) [reply pkg/log/status.go]: Ah, so the "is terminal" check, if true, will use print using a carriage return and the output gets jumbled up with the logs.

File: pkg/log/status.go [hunk 3]
```diff
@@ -103,14 +103,8 @@ func IsTerminal(w io.Writer) bool {
 func (s *Status) Start(status string) {
 	s.End(true)
 	// set new status
-	isTerm := IsTerminal(s.writer)
```
[the remark this answers is not in the document]
Comment (alejandrox1) [reply pkg/log/status.go]: Ah, so the "is terminal" check, if true, will use print using a carriage return and the output gets jumbled up with the logs.

File: Gopkg.toml [hunk 4]
```diff
@@ -55,3 +56,6 @@ required = [
  name = "k8s.io/code-generator"
  branch = "release-1.11"
 
+[[constraint]]
+  name = "k8s.io/klog"
+  version = "0.1.0"
```
[the remark this answers is not in the document]
Comment (alejandrox1) [reply Gopkg.toml]: I don't particularly think so but I wanted to cover all my bases.

File: cmd/kind/kind.go (hunk 1 above)
Comment (munnerz) [reply cmd/kind/kind.go]: I don't think we can explicitly add our own flags for this - instead, we should be calling `klog.InitFlags` with the cobra command's flagset: https://github.com/kubernetes/klog/blob/master/klog.go#L406-L420

Comment (alejandrox1) [reply cmd/kind/kind.go]: Forgot about `InitFlags`. Let me add it

File: pkg/cluster/cluster.go [hunk 5]
```diff
@@ -106,7 +106,6 @@ func (c *Context) Create(cfg *config.Config) error {
 
 	fmt.Printf("Creating cluster '%s' ...\n", c.ClusterName())
 	c.status = logutil.NewStatus(os.Stdout)
-	c.status.MaybeWrapLogrus(log.StandardLogger())
```
[the remark this answers is not in the document]
Comment (alejandrox1) [reply pkg/cluster/cluster.go]: Alright. I wanted to see what yall thought about this part. So let me work on something and i'll see how to add back in the "write to terminal" parts from the previous comments.

File: cmd/kind/kind.go (hunk 1 above)
Comment (BenTheElder) [reply cmd/kind/kind.go]: We can plumb through only the flags we want. At minimum v, the others are more detabateble.

File: pkg/log/status.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenTheElder) [reply pkg/log/status.go]: the carriage return is for the spinner, previously these worked together by way of the output wrapper.

this means that if it's not in a terminal, it won't spin, but if it is, you get the spinner while waiting on various things

File: Gopkg.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenTheElder) [reply Gopkg.toml]: let's unpin it so we can freely upgrade

File: cmd/kind/kind.go [hunk 6]
```diff
@@ -67,38 +53,20 @@ func NewCommand() *cobra.Command {
...
 	return NewCommand().Execute()
 }
 
+func init() {
+	//log.InitFlags(nil)
+	flag.CommandLine.AddGoFlagSet(goflag.CommandLine)
+}
+
 // Main wraps Run, adding a log.Fatal(err) on error, and setting the log formatter
 func Main() {
```
Comment (BenTheElder) [inline cmd/kind/kind.go]: we should set up klog here

File: cmd/kind/kind.go [hunk 7]
```diff
@@ -67,38 +53,20 @@ func NewCommand() *cobra.Command {
...
-	// particularly useful, they're relative to the command start
-	log.SetFormatter(&log.TextFormatter{
-		FullTimestamp:   true,
-		TimestampFormat: "15:04:05",
-		// we force colors because this only forces over the isTerminal check
-		// and this will not be accurately checkable later on when we wrap
-		// the logger output with our logutil.StatusFriendlyWriter
-		ForceColors: logutil.IsTerminal(log.StandardLogger().Out),
-	})
 	if err := Run(); err != nil {
```
Comment (BenTheElder) [inline cmd/kind/kind.go]: we'll probably have to make this:
```
cmd := NewCommand()
// setup klog
klogFlags := golfag.NewFlagSet("kind", goflag. ContinueOnError)
log.InitFlags(klogFlags)
cmd.PersistentFlags().AddGoFlagSet(klogFlags)
// now we can run
if err := cmd.Execute(); err != nil {
```

Comment (BenTheElder) [reply cmd/kind/kind.go]: which is a little unfortunate. at some point ideally we can bind klog flags in NewCommand at the root instead.

File: pkg/log/status.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alejandrox1) [reply pkg/log/status.go]: Fixed this as well.

File: Gopkg.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (alejandrox1) [reply Gopkg.toml]: unpinned

**Comment by BenTheElder** [comment]
(I just rebased it [using this](https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/))

**Comment by BenTheElder** [comment]
tinkering with this locally there's a few more changes to make everything play nice, I expect to get this change in today though
/hold

**Comment by BenTheElder** [comment]
I did some more work on this, getting everything to play nice is super painful :(

I think maybe instead I'll set up a log-shim and support forwading the shim to klog...

**Comment by BenTheElder** [comment]
small update: discussing some options around this with upstream klog developers :-)
will get back to this soon!

**Comment by alejandrox1** [comment]
@BenTheElder @neolit123 should I close this PR? It looks like it is way out of date and not very useful given the current state of the repo

**Comment by BenTheElder** [comment]
yeah, I was hopeful we could switch but I think it's probably not ready
yet, and if it is we'll need a significant update to bring this back up to
speed, let's just come back to this later

thanks again anyhow :-)

**Pull request closed without merging**